### ISSUE TITLE
How inter-dependent are the DPR and the reader?

### ISSUE THREAD

**Issue by morningmoni** [facebookresearch/DPR#18]
Greetings,

I am wondering how the performance of the reader depends on the retriever. 

1. In the paper Sec 6.1, it says "For training, we sample one positive and m− 1 negative passages for each question at each iteration".  Is this one positive passage retrieved by DPR or any passage (described in Sec 4.2 Selecting positive passages) with has_answer=True will do? Or rather, is the reader training independent of the retriever training?

2. In the same paragraph, it says "we use the highest-ranked passage from BM25 that contains the answer as the positive passage". Do you use A or Q+A as the query for BM25?

3. I assume the order of retrieved passages doesn't matter for the reader as its probability is normalized by softmax of all retrieved passages. Is this correct?

Many thanks!

**Comment by vlad-karpukhin** [comment]
Hello!

Since DPR is e2e pipeline that means reader depends on what is returned from the retriever component, for both training and inference.

1.  Yes, positive passage is sampled from the results of the dense retriever component. But there are some extra heuristics we used for compose reader training data and omitted in the paper for simplicity. 
We used NQ gold wikipedia page references for positive passage selections - whenever there are retriever results from the 'gold' page, we prefer those vs positive from non gold pages, otherwise just use top k best positives.
You can have a look at this logic and different parameters in the code at:
https://github.com/facebookresearch/DPR/blob/master/dpr/data/reader_data.py#L90
https://github.com/facebookresearch/DPR/blob/master/dpr/data/reader_data.py#L103

2. Sorry, there is no "we use the highest-ranked ... " text in that paragraph in section #6.1. We used bm25 results only to mine hard negative samples for the retriever training. BTW, you can use DPR to mine hard negatives and then fine-tune DPR again - this gives even better results reported in the paper.
We've never used Q+A as a query for bm25.

3. We used top 100 or 200 results form the retriever to compose the pools of positive and negative passages for each question using the logic referenced in #1 above. Yes, the reader's model ranker component normalizes the scores but which passages will be selected in the pools still has some dependencies on its retrieval order. 
For example, we take up to k first positive results form the retriever. If a positive passage is not that list, it won't participate in the reader model training process.

Hope this answers your questions

**Comment by morningmoni** [comment]
Thanks a lot for your detailed reply!

I have another question regarding this. When using DPR+BM25 for retrieval, do you re-train a reader using the retrieval results of DPR+BM25 or simply conduct inference of the same reader trained using DPR retrieval on the newly retrieved passages? How do you think the performance would differ (since the distributions of the two may be different)?

For Q2, "we use the highest-ranked passage from BM25 that contains the answer as the positive passage." is in Sec 4.2.

**Comment by vlad-karpukhin** [comment]
Hello there,

"When using DPR+BM25 for retrieval, do you re-train a reader using the
retrieval results of DPR+BM25 or simply conduct inference of the same
reader trained using DPR retrieval on the newly retrieved passages? " - We
use corresponding retrieval results for every reader. I.e. DPR+BM25 results
were used to train the reader if the table's model column says so, likewise
DPR only results were used to train the reader when column "Model"=DPR. For
NQ dataset and the reader training, we additionally used that gold
wikipedia page links for extra positive passages selection logic I
explained earlier.

"we use the highest-ranked passage from BM25 that contains the answer as
the positive passage." - this is about selecting positive passages for the
datasets where gold contexts are not provided.

Thanks,
Vlad
